src/Common/src/Interop/Windows/HttpApi/Interop.HttpApi.cs

```diff
@@ -0,0 +1,1021 @@
+                if (result.IsInvalid)
+                {
+                    result.SetHandleAsInvalid();
+                    throw new OutOfMemoryException();
+                }
+
+                result._size = cb;
+                return result;
+            }
+
+            override protected bool ReleaseHandle()
+            {
+                return Interop.mincore_obsolete.LocalFree(handle) == IntPtr.Zero;
+            }
+        }
+
+        [DllImport(Interop.Libraries.Heap, EntryPoint = "LocalAlloc", SetLastError = true)]
+        internal static extern SafeLocalFreeChannelBinding LocalAllocChannelBinding(int uFlags, UIntPtr sizetdwBytes);
+
+        [DllImport(Libraries.HttpApi, SetLastError = true)]
+        internal static extern uint HttpReceiveClientCertificate(SafeHandle requestQueueHandle, ulong connectionId, uint flags, HTTP_SSL_CLIENT_CERT_INFO* pSslClientCertInfo, uint sslClientCertInfoSize, uint* pBytesReceived, NativeOverlapped* pOverlapped);
+
+        [DllImport(Libraries.HttpApi, SetLastError = true)]
+        internal static extern uint HttpReceiveClientCertificate(SafeHandle requestQueueHandle, ulong connectionId, uint flags, byte* pSslClientCertInfo, uint sslClientCertInfoSize, uint* pBytesReceived, NativeOverlapped* pOverlapped);
+
+        //
+        // TODO: move this to a better location
+        //
+
+        [Flags]
+        internal enum FileCompletionNotificationModes : byte
+        {
+            None = 0,
+            SkipCompletionPortOnSuccess = 1,
+            SkipSetEventOnHandle = 2
+        }
+
+        [DllImport(Libraries.Kernel32_L2, SetLastError = true)]
+        internal static unsafe extern bool SetFileCompletionNotificationModes(SafeHandle handle, FileCompletionNotificationModes modes);
+
+        internal static readonly string[] HttpVerbs = new string[]
+        {
+            null,
+            "Unknown",
+            "Invalid",
+            "OPTIONS",
+            "GET",
+            "HEAD",
+            "POST",
+            "PUT",
+            "DELETE",
+            "TRACE",
+            "CONNECT",
+            "TRACK",
+            "MOVE",
+            "COPY",
+            "PROPFIND",
+            "PROPPATCH",
+            "MKCOL",
+            "LOCK",
+            "UNLOCK",
+            "SEARCH",
+        };
+
+        internal static class HTTP_RESPONSE_HEADER_ID
+        {
+            private static Dictionary<string, int> s_hashtable;
+
+            static HTTP_RESPONSE_HEADER_ID()
+            {
+                s_hashtable = new Dictionary<string, int>((int)Enum.HttpHeaderResponseMaximum);
+                for (int i = 0; i < (int)Enum.HttpHeaderResponseMaximum; i++)
+                {
+                    s_hashtable.Add(s_strings[i], i);
+                }
+            }
+
+            internal static int IndexOfKnownHeader(string HeaderName)
+            {
+                object index = s_hashtable[HeaderName];
+                return index == null ? -1 : (int)index;
```

Comment:
Was this code originally using a Hashtable instead of a Dictionary?  Seems like this could blow up with an exception if the HeaderName doesn't exist, and should instead use TryGetValue, which also avoids the boxing on each call.